Add support for Arturia KeyLab Mk1

SamWhited opened the pull request:
The corresponding manual PR is here:

https://github.com/mixxxdj/manual/pull/744

![Graphical depiction of the keylab to give a sense of where the controls are. A textual description is included in the readme for the downstream project and will be included in the Mixxx manual.](https://github.com/user-attachments/assets/eca72d68-a64d-4c4a-a5f2-e4c7fa51cd56)


A few things to note in particular:

  - The menu system is a bit complicated. I don't see a way around that, but maybe someone else will have a better idea.
  - Unsure what's actually useful to have in the on-screen menu or not. I threw in a few actions, but some may or may not be as useful as others and I'm open to tweaking those.
  - Sound and Multi feel weird as deck switching buttons, but nothing else really worked well.
  - Fader bank 2 is unused at the moment, so there's room to make a more comprehensive effects controlling system, but that may be for a later update to keep the complexity of this PR low. Open to ideas though.

SamWhited committed:
Add support for Arturia KeyLab Mk1

Signed-off-by: Sam Whited <sam@samwhited.com>

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,759 @@
+    };
+
+    class Deck extends components.Deck {
+        currentDeckNum() {
+            const deckStr = script.channelRegEx.exec(this.currentDeck)[1];
+            return parseInt(deckStr);
+        }
```
`script.deckFromGroup`?

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,759 @@
+                    type: components.Button.prototype.types.push,
+                    blinkState: false,
+                    blinkTimer: 0,
+                    output: function(value) {
+                        if (value === 2) {
+                            if (this.blinkTimer === 0) {
+                                this.blinkTimer = engine.beginTimer(250, () => {
+                                    this.blinkState = !this.blinkState;
+                                    delayLED(0x12 + i)(this.blinkState ? 1 : 0);
+                                });
+                            }
+                        } else {
+                            if (this.blinkTimer) {
+                                engine.stopTimer(this.blinkTimer);
+                                this.blinkTimer = 0;
+                            }
+                            delayLED(0x12 + i)(value);
+                        }
```
can you elaborate what this does/why its needed?

SamWhited replied:
This blinks the button for the currently active loop cue so you can keep track of where in the song you are. It's not super important, but I found it to be a nice touch so that I know what loop is active and can activate the next one, jump to a previous one, or deactivate a loop without accidentally getting the wrong button.

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,759 @@
+                    if (!this.inKey) {
+                        return 0;
+                    }
+                    return engine.getValue(this.group, this.inKey.replace("[Channel1]", this.activeDeck.currentDeck));
```
why this workaround?

SamWhited replied:
Because the key is being set from the menu system it's not getting correctly changed when we swap the deck over. This workaround makes sure it gets changed even if we change the value of inKey after the deck has been selected.

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,759 @@
+                        this.activeDeck.valueEncoder.group = this.activeDeck.currentDeck;
+                        this.activeDeck.valueButton.group = this.activeDeck.currentDeck;
+                    }
+                    this.activeDeck.valueEncoder.name = param.name;
+                    this.activeDeck.valueButton.name = param.name;
+                    this.activeDeck.valueEncoder.inKey = param.inKey;
+                    this.activeDeck.valueEncoder.max = param.max;
+                    this.activeDeck.valueEncoder.min = param.min;
+                    this.activeDeck.valueEncoder.relative = param.relative || false;
+                    this.activeDeck.valueButton.inKey = param.pressInKey;
+                    this.activeDeck.valueButton.type = param.pressType || components.Button.prototype.types.push;
+                    this.activeDeck.valueEncoder.outputValue = param.outputValue;
+                    this.activeDeck.valueButton.outputValue = param.outputValue;
```
`Object.assign` may be useful to clean this up.

SamWhited replied:
I thought of that, but since the key isn't always the same name as what Mixxx expects (ie. "pressInKey" for the buttons "inKey") that doesn't work. Maybe I should swap the menu over to have separate "button" and "encoder" objects that have the expected key names so I can just copy that whole object instead. Although it still feels odd to not explicitly show what's supported. I'll play with it and see what works. Thanks.

SamWhited replied:
Done, this is much cleaner looking now. I also noticed a minor bug where the name wasn't being set so it showed up as "Undefined" on the menu when, so eg. when sorting columns it would say "Undefined: artist" instead of "Sort: artist". I also removed some unnecessary copying, the menu slice is never modified so we can just reference it directly.

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,759 @@
+                    number: i + 1,
+                    midi: [0x99, pad(i)], // Drum pads.
+                    volumeByVelocity: true,
+                    // These numbers are made up magic constants that can be
+                    // checked in the "send" callback defined below.
```
who made them up? Or rather, have you defined them or do they come from the firmware?

SamWhited replied:
This is all built into componentsJS. Send gets called with the value of `this.playing` or `this.empty` etc. for the purpose of setting the color. We can't set the color, so I just made up some random magic numbers and defined them here so that we'd have something to check in the switch function so we'd know what action is being sent.

Swiftb0y reviewed (commented):
couple thoughts. in good shape already

SamWhited committed:
keylab: address review feedback

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
KeyLab: update stop button to be the cue button

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Update: changed the stop button to be a cue button, which is more useful IMO.

Also made the corresponding manual PR: https://github.com/mixxxdj/manual/pull/744

Really the last thing to do is figure out a good layout for the snapshot buttons; I think that instead of cues they should probably toggle between different modes for the pads. Suggestions for how this could best be handled or configured recommended.

SamWhited committed:
KeyLab: cleanup setting menu options

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Document KeyLab USB details in XML file

Also add a link to the manual

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
KeyLab: rework drum pads to be configurable

And add some basic configurations for samplers, hotcues, and intro/outro
markers + hotcues.

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Just pushed up a major change to make the drum pads work like the ones on the Traktor S4 Mk3. This lets you change the default pad configuration between "intro/outro cue + hotcues", "hotcues", or "samplers". The idea is that the first four snapshot buttons will change to quickly toggle between these modes.

For some reason the `applyLayer` method of component containers isn't working unless I copy it out of the parent class into the subclass I made. It's unclear to me why this is, something about JavaScript's this binding I assume.

SamWhited committed:
KeyLab: add buttons to switch between pad modes

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
This is probably ready for review again. I've added support for switching between different pad modes, though I'm not very happy with how it turned out and everything feels a bit jank (though it does work). I'll think about it more this weekend and hopefully come up with a better way to handle this, or at least deduplicate some of the code.

SamWhited committed:
KeyLab: only show layer on screen on button press

Previously this would be written any time the layer was updated, eg.
when starting the controller for the first time, resulting in the screen
randomly displaying "Deck 1".

Signed-off-by: Sam Whited <sam@samwhited.com>

Swiftb0y commented:
Thanks for letting me know. I have a lot on my plate right now so I can't give this a proper review atm. please ping me on the 5th of April again or find another reviewer. Thank you.

SamWhited commented:
No pressure, of course! If anyone else has the time, I think this is probably pretty close to ready.

SamWhited committed:
KeyLab: factor out the layer mapping code

Also remove some unnecessary "TODO" comments.

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Gentle ping now that it's after the 5th; no rush of course. Thanks!

SamWhited commented:
@Swiftb0y gentle ping again; no worries if you're still swamped, I just wanted to make sure this didn't slowly fall off the radar. Thanks!

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,1007 @@
+
+    KeyLabMk1.incomingData = function(data, length) {
+        if (length < 6) {
+            console.log(`expected sysex packet of length 6, got ${length}`);
+            return;
+        }
+        for (let n = 0; n < MMCHeader.length; n++) {
+            if (data[n] !== MMCHeader[n]) {
+                console.log("unknown sysex packet");
```
NIT: consider logging the packet too. that way the logs would be more useful by default.

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,1007 @@
+                        return;
+                    }
+                    const param = this.inGetParameter();
+                    if (value === 0x3F) {
+                        if (this.relative) {
+                            this.inSetParameter(-1);
+                            return;
+                        }
+                        const newParam = param - 1;
+                        if (
+                            this.min !== undefined && this.max !== undefined
+                            && newParam < this.min
+                        ) {
+                            this.inSetParameter(this.max);
+                        } else {
+                            this.inSetParameter(param - 1);
+                        }
+                    } else if (value === 0x41) {
+                        if (this.relative) {
+                            this.inSetParameter(1);
+                            return;
+                        }
+                        const newParam = param + 1;
+                        if (
+                            this.max !== undefined && this.min !== undefined
+                            && newParam > this.max
+                        ) {
+                            this.inSetParameter(this.min);
+                        } else {
+                            this.inSetParameter(param + 1);
+                        }
+                    }
```
this looks very convoluted. I also think there is a bug in the clamping logic (if `newParam > this.max` but then you `this.inSetParameter(this.min);` and vice versa). 
```suggestion
                    value = value - 0x40;
                    if (this.relative) {
                        this.inSetParameter(value);
                        return;
                    }
                    const newParam = param + value;
                    if (this.min !== undefined && this.max !== undefined) {
                        this.inSetParameter(script.clamp(newParam, this.min, this.max));
                    } else {
                        this.inSetParamter(newParam);
                    }
```
for this to work, add a `script.clamp = function(val, low, high) {return Math.min(high, Math.max(low, val));}` to `common-controller-scripts.js`.

SamWhited replied:
I had forgotten about this; I didn't realize it was just 2's compliment values that are being used to indicate velocity when I did this. I've since written a few more and realized why it's 0x41 or what not. I'll change this over to the regular implementation.

It might be worth having a simple way to decode and scale the value in the `script` namespace too (that way those writing XML mappings have access to it for infinite encoders, which right now just aren't supported in XML-only stuff).

SamWhited replied:
I did make some changes to make this work regardless of how fast you turn the encoder, but the logic ended up being basically the same. I had forgotten that this isn't clamping, it's wrapping around on the ends between min and max (ie. if we sort by library columns and we hit last_played (31) we want to sort by artist (1) next, not an undefined value). I suppose we could add some sort of wrap function too, but it seems less useful so I left the logic as-is. I'm still open to simplification, of course, but this seems the most readable way for now.

Swiftb0y commented on res/controllers/Arturia-KeyLab-Mk1-scripts.js:
```diff
@@ -0,0 +1,1007 @@
+        rotate: {
+            inKey: "key",
+            max: 24,
+            outputValue: (value) => {
+                return keyNums[script.posMod(value - 1, keyNums.length)];
+            },
```
isn't this just `outValueScale`?

SamWhited replied:
Oops, yup, that's basically doing the same thing. Fixed.

Swiftb0y reviewed (commented):
couple more thoughts. Don't have time for more right now.

SamWhited committed:
KeyLab: fixed a few issues from review

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Update logic on infinite encoder scroll

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Rebased now that 2.5.1 is out. It would be great to get another review if you have time!

daschuer commented:
@Swiftb0y Are all you comments addressed?

SamWhited commented:
Gentle ping.

Swiftb0y reviewed (approved):
Nothing obvious, no. Sorry for the delay.
